### PR TITLE
Merge CropUI with v1.96_dev branch

### PR DESCRIPTION
**Pull request by Nenotriple**
CropUI can be opened by starting img-txt Viewer, selecting a directory, and then using the menubar or image context menu command `Crop...`

---

# Todo:
#### General
- [x] Implement CropUI into the ImgTxtViewer UI
- [x] Logic - Files should be sorted based on the sort key returned by `ImgTxtViewer.get_file_sort_key()`
- 
- [x] UI - Option to disable the highlight/dark background during selection.
  - [x] - Logic
- [x] UI - Display number of images in directory, and current index.
  - [x] - Logic
- [x] UI - Display selection aspect ratio in `info_label`
  - [x] - Logic
- [x] UI - Double-click to automatically create a selection.
- [x] UI - Image transform controls - Rotate / Flip
  - [x] - Logic

#### After Cropping
- [x] UI - Save & Close - Save cropped image and close CropUI.
  - [x] - Logic
- [x] UI - Save & Next - Save the cropped image and move to the next index.
  - [x] - Logic
- [x] UI - Save As... - Ask for filename/path when saving, don't close CropUI, don't advance index.
  - [x] - Logic
- [x] UI - Save - Save cropped image, don't close CropUI, don't advance index.
  - [x] - Logic
- [X] - UI Overwrite - Overwrite the current image with the cropped image.
  - [x] - Logic
- 
- [ ] UI - *(Optional)* An option to draw composition guides.
- [ ] UI - *(Optional)* Resize output to selection size, with options: `Upscale`, `Downscale`, `Up/Downscale`
- [ ] UI - *(Optional)* Display all frames of a GIF in a thumbnail panel, allow selecting a frame to crop it.

---

#### Crop Selection:
- [x] UI - Automatic selection:
  - [x] - Logic - Auto-size selection to the closest matching ratio from predefined ratios based on the input image ratio.
- [x] Logic - `fixed_selection_var` - Enable or disable fixed selection options
  - [x] UI
- [x] Logic - `fixed_selection_option_var` - Fixed selection options - `"Aspect Ratio", "Width", "Height", "Size"`
  - [x] UI
- [x] Logic - `fixed_selection_entry_var` - Adjust the parameters for the selected `fixed_selection_option_var`.
  - [x] UI
- [x] Enforce fixed selection when using handles.
- [x] Ensure `expand from center`, and `fixed selection` work together.
-
- [x] UI - A button that inserts the aspect ratio or size of the current selection as the `fixed` value.
  - [x] - Logic
- 
- [x] Logic - `expand_from_center_var` - Expand equally from center instead of top-left corner.
  - [X] Normal selection
  - [x] When grabbing a corner handle, the selection expands equally from the center.
  - [x] When grabbing a side handle, the selection expands equally from the opposite side.

---
#### Issues:
- [x] Fixed modes, Width and Height: Input values > image dimensions cause selection overrun.
- [x] Very small images (?) aren't being cropped correctly.
- [x] image scale dictates crop selection accuracy.
  - [x] separate image coords from canvas coords.
- [x] Maximizing and then un-maximizing incorrectly displays the image, and offsets.
- [x] Logic - `Position` x/y entries are not working correctly and need to be fixed.

---

### CODE REVIEW

**Comment by UrzasLegacy** [comment]
So I have an idea about this, lemme know what you think.

If we define our use case, we are trying to easily crop images into ratio buckets, right? I can't think of a reason I'd want to crop to any arbitrary resolution; I could just use my usual editing software for that.

That said, you could make bucket customization options like minimum bucket resolution, maximum bucket resolution, and bucket resolution step and populate a list of options from that. Once a bucket ratio is selected, rather than manipulating a selection box, a rectangle of the chosen bucket resolution follows the mouse and the user can just align that box on the image and a single click would save a cropped copy of the image. Like a cookie cutter.

Additionally, I try to keep my images as high res as possible (to a point lol) so I focus on the aspect ratios with the goal of preserving as many pixels as I can instead of aiming for specific resolutions. I just let Kohya handle the downsizing part.

Ideally it'd work something like this:
- Automatically determine the best bucket aspect ratio for the given image's resolution maximizing for pixel area.
- Apply that ratio to the current image resolution to get a template resolution.
- Determine current scalar of the image as displayed in the app (zoomed in/out) and adjust template resolution/position to match,
- Draw a rectangle of that template resolution that centers on and follows the cursor.
- When user clicks on the image, map the template resolution/position back to the actual image resolution/position.
- Save a copied, cropped image to an intermediary folder possibly with subfolders for each aspect ratio

While training bbox detection models, I was looking for a solution like this. I did find some projects that use the cookie cutter approach, but nothing that does the automatic bucketing and really ties it all together with a nice UI.

The functionality of this one is nice and it's implemented in tkinter, but it's painful to use:
https://github.com/theallyprompts/PixelPruner

Another example, though it uses PySide for UI and it's very rough. It's neat though because it lets you crop multiple different resolutions in one click:
https://github.com/Anzhc/Anzhc-s-Ez-Cropper/blob/main/main.py

**Comment by Nenotriple** [comment]
The general scope of the tool was to replicate the basic cropping functions that are available in GIMP.

> we are trying to easily crop images into ratio buckets, right? I can't think of a reason I'd want to crop to any arbitrary resolution

This isn't exactly my use case. I crop images to improve context, so it's not always about preserving the most pixels. When I crop a dataset, I generally use 3 aspect ratios: `1:1`, `4:3`, and `4:5`, while ensuring the crop resolution is at or above my training resolution, but not larger than 2048px. Then I just let the training script bucket or resize however it handles things.

> Ideally it'd work something like this:

I think some automatic modes would be a good idea. Once all of the fundamental selection logic is done, it should be pretty simple to manipulate the crop selection automatically.

> Automatically determine the best bucket aspect ratio for the given image's resolution maximizing for pixel area.

I don't really grasp what this gains: `best aspect ratio... maximizing for pixel area`, If we try for the optimal aspect ratio to cover the most pixels, that just sounds like the original image? I think I need a visual aid for this like a toddler, cause I don't really understand.

---

I'll take a look at those apps, see if there's anything I can stea- *borrow* :)

**Comment by UrzasLegacy** [comment]
Okay, so bear with me 'cause I'm not a master at dataset prep or anything and I'm somewhat new to lora training lol. I could very well be completely overthinking this.

> I crop images to improve context, so it's not always about preserving the most pixels.

You're absolutely right, context is the important part, but the end goal is still to have an image with context framed as you'd like and that fits into one of your chosen aspect buckets. I use `1:1`, `3:4`, `5:8`, `4:3`, `3:2`, `16:9`.

The preserving pixels part isn't about the actual training so much, but more just about maintaining as much fidelity as possible in the original dataset for manual processing purposes. This isn't always a priority and you can always upscale, but for certain source images of already high quality resolution, it made sense to me to consider pixel preservation when cropping.

The way I'm doing it, I have to sorta eyeball what I think my end aspect ratio will be, set that ratio scalar in Krita for my crop tool, lock it, and then drag handles to resize it and move it around.

Instead, it'd be nice to be able to auto-detect the closest aspect ratio, create a selection rectangle on the mouse with a resolution close to the original image resolution and then allow resizing with the scroll wheel while maintaining that aspect ratio. Heck, shift+scroll wheel to flip through aspect ratios would be badass too :)

**Comment by Nenotriple** [comment]
I think I understand now:

Define a list of ratios like `1:1, 3:4, 5:8, 4:3, 3:2, 16:9`, When creating a selection, it will automatically be sized to the listed ratio that most closely matches the input image ratio.

Is that right?

**Comment by UrzasLegacy** [comment]
Yup, you got it. Throw in some hotkeys to easily cycle through images, adjust image size, template size, aspect ratio, image rotation, etc and you're cookin.

File: CropUI.py
Comment (Nenotriple) [inline CropUI.py:1]: Introduced tooltips for various UI elements.

Modified selection and resizing logic for image cropping.

Added new methods for expanding from the center and clamping coordinates to the image.

Updated the UI layout and added new control panel elements.

Adjusted image display and cropping functionalities.

**Pull request merged**